allennlp/training/trainer2.py

```diff
@@ -0,0 +1,673 @@
+                               'meaning that early stopping is disabled')
+        elif (not isinstance(patience, int)) or patience <= 0:
+            raise ConfigurationError('{} is an invalid value for "patience": it must be a positive integer '
+                                     'or None (if you want to disable early stopping)'.format(patience))
+
+        tensorboard = TensorboardWriter(
+                get_batch_num_total=lambda: self.batch_num_total,
+                serialization_dir=serialization_dir,
+                summary_interval=summary_interval,
+                histogram_interval=histogram_interval,
+                should_log_parameter_statistics=should_log_parameter_statistics,
+                should_log_learning_rate=should_log_learning_rate)
+
+        # Enable activation logging.
+        if histogram_interval is not None:
+            tensorboard.enable_activation_logging(model)
+
+        # This is all state that the callbacks might want:
+        # I am not calling move_to_gpu here, because if the model is
+        # not already on the GPU then the optimizer is going to be wrong.
+        self.model = model
+        self.iterator = iterator
+        self.validation_iterator = validation_iterator or iterator
+        self.shuffle = shuffle
+        self.optimizer = optimizer
+        self.train_data = train_dataset
+        self.validation_data = validation_dataset
+
+        # For capturing mid / end-of-epoch metrics
+        self.train_metrics: Dict[str, float] = {}
+        self.val_metrics: Dict[str, float] = {}
+
+        # For capturing overall metrics
+        self.metrics: Dict[str, Any] = {}
+
+        self.batch_num_total = 0
+        self.cumulative_batch_size = 0
+        self.batch_group = None
+        self.batches_this_epoch = 0
+        # For tracking is_best_so_far and should_stop_early
+        self.metric_tracker = MetricTracker(patience, validation_metric)
+        # Get rid of + or -
+        self.validation_metric = validation_metric[1:]
+        self.num_epochs = num_epochs
+
+        self.checkpointer = checkpointer
+        self.checkpoint_epoch: Union[int, str] = 0
+        self.model_save_interval = model_save_interval
+
+        self.grad_norm = grad_norm
+        self.grad_clipping = grad_clipping
+        self.learning_rate_scheduler = learning_rate_scheduler
+        self.momentum_scheduler = momentum_scheduler
+        self.moving_average = moving_average
+        self.tensorboard = tensorboard
+        self.last_log = 0.0
+        self.epoch_number = 0
+        self.batch_grad_norm: Optional[float] = None
+
+        # Set up callback handler
+        callbacks: List[Callback] = [
+                LogTensorboard(log_batch_size_period),
+                LogTensorboardHistograms(),
+                LrsCallback(),
+                CheckpointCallback(),
+                MovingAverageCallback(),
+                Validate()
+        ]
+
+        self.handler = CallbackHandler(callbacks, self)
```

Comment:
So, I'm not sure what we get by using `Protocols` instead of just typing all of the `State` objects that you have just as `Trainer`. You (might) gain some flexibility in how you use the callbacks, but it comes at a pretty high cognitive cost, as @DeNeutoy points out.  I'd vote for just using `Trainer` everywhere, and pushing as much of the necessary state into the constructor parameters of the individual callbacks as possible.